File: operators/compute/basic/embedding.py
# Copyright (c) 2024 The Core Team
#
# Licensed under the Apache License, Version 2.0
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import torch
from typing import Tuple, Optional, List
import math

from . import init

def embedding_init(
    num_embeddings: int, embedding_dim: int, padding_idx: Optional[int] = None,
    max_norm: Optional[float] = None, norm_type: float = 2., scale_grad_by_freq: bool = False,
    sparse: bool = False, device: torch.device = 'cuda:0', dtype=torch.float32, training=True,
):
    if padding_idx is not None:
        if padding_idx > 0:
            assert padding_idx < num_embeddings, 'Padding_idx must be within num_embeddings'
        elif padding_idx < 0:
            assert padding_idx >= -num_embeddings, 'Padding_idx must be within num_embeddings'
            padding_idx = num_embeddings + padding_idx
    weight = torch.zeros((num_embeddings, embedding_dim)).to(device=device, dtype=dtype)
    weight = init.normal_(weight)
    weight = _fill_padding_idx_with_zero(weight, padding_idx)
    init_info = (max_norm, norm_type, scale_grad_by_freq, sparse)
    if training:
        weight.grad = torch.zeros_like(weight, device=weight.device, dtype=weight.dtype)
    return weight, init_info

def embedding_forward(input: torch.Tensor, weight: torch.Tensor, init_info: Tuple):
    # if input.dim() < 2:
    #     raise ValueError("Input tensor must be at least 2D")
    max_norm, norm_type, scale_grad_by_freq, sparse = init_info
    if max_norm is not None:
        # Note [embedding_renorm contiguous]
        # `embedding_renorm_` will call .contiguous() on input anyways, so we
        # call it here and take advantage of the improved locality in the
        # `embedding` call below too.
        input = input.contiguous()
        # Note [embedding_renorm set_grad_enabled]
        # XXX: equivalent to
        # with torch.no_grad():
        #   torch.embedding_renorm_
        # remove once script supports set_grad_enabled
        _no_grad_embedding_renorm_(weight, input, max_norm, norm_type)
    save_for_backward = (input)
    input_size = list(input.size())
    output = weight.index_select(0, input.view(-1)).view((*input_size, weight.size(-1)))
    return output, save_for_backward

def embedding_backward(grad_output: torch.Tensor, weight: torch.Tensor, save_for_backward: Tuple[torch.Tensor, torch.Tensor]):
    input = save_for_backward
    grad_input = None
    grad_weight = torch.zeros_like(weight, device=weight.device, dtype=weight.dtype)
    # for i in range(input.numel()):
    #     index = input.view(-1)[i]
    #     grad_weight[index] += grad_output.view(-1, weight.size(1))[i]
    input_flat = input.view(-1)
    grad_output_flat = grad_output.view(-1, weight.size(1))
    grad_weight.index_add_(0, input_flat, grad_output_flat)
    weight.grad += grad_weight
    return grad_input, weight


def _fill_padding_idx_with_zero(weight: torch.Tensor, padding_idx) -> None:
    if padding_idx is not None:
        weight[padding_idx].fill_(0)
    return weight

def _no_grad_embedding_renorm_(weight: torch.Tensor, input: torch.Tensor, max_norm: float, norm_type: float) -> Tuple[torch.Tensor, torch.Tensor]:
    torch.embedding_renorm_(weight, input, max_norm, norm_type)

File: src/compute/embedding.cpp
/*
 * Copyright (c) 2024 The Core team
 *
 * Licensed under the Apache License, Version 2.0;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Header Order Protection
// ReSharper disable once CppUnusedIncludeDirective
#include <c10/util/Exception.h>
// Header Order Protection

#include <torch/nn/functional/embedding.h>

#include "compute/embedding.h"
#include "logger.h"
#include "tensor_impl.h"
#include "threading/scheduler_impl.h"
#include "threading/task_impl.h"

namespace cs::compute {
Embedding::State::State(const Forward &forward, const Backward &backward,
                        const Args &args)
    : forward{forward}, backward{backward}, args{args} {}

OrderedDict<std::string, Tensor> Embedding::State::parameters() const {
  OrderedDict<std::string, Tensor> dict;
  dict.insert("weight", forward.weight);
  return dict;
}

OrderedDict<std::string, Tensor> Embedding::State::gradients() const {
  OrderedDict<std::string, Tensor> dict;
  dict.insert("weight", forward.grad_weight);
  return dict;
}

OrderedDict<std::string, module::State::Increment>
Embedding::State::increments() const {
  OrderedDict<std::string, Increment> dict;
  dict.insert("weight", {forward.weight, forward.grad_weight});
  return dict;
}

void Embedding::State::zero_grad() { forward.grad_weight = {}; }

std::shared_ptr<Embedding::State> Embedding::init(const Scheduler &scheduler,
                                                  const Options &options) {
  struct Impl : Task::Impl {
    const Options options;

    explicit Impl(std::vector<Tensor> output /* weight */,
                  const Options &options)
        : Task::Impl{std::move(output), {}, kMain, kCompute},
          options{options} {}
    void operator()() const override {
      const auto weight_ =
          at::normal(0, 1, {options.num_embeddings(), options.embedding_dim()},
                     {}, options.dtype(), {}, options.device(), {});
      // if (max_norm != c10::nullopt) {
      //   input_ = input_.contiguous();
      //   _no_grad_embedding_renorm_(weight, input_, *max_norm, norm_type);
      // }
      output()[0].impl()->tensor() = weight_;
    }
    [[nodiscard]] const char *name() const override {
      return "cs::compute::Embedding::init";
    }
  };

  int64_t padding_idx = -1;
  if (options.padding_idx() != c10::nullopt) {
    if (*options.padding_idx() > 0) {
      TORCH_CHECK(*options.padding_idx() < options.num_embeddings(),
                  "Padding_idx must be within num_embeddings");
    } else if (*options.padding_idx() < 0) {
      TORCH_CHECK(*options.padding_idx() >= -options.num_embeddings(),
                  "Padding_idx must be within num_embedding");
      padding_idx = options.num_embeddings() + *options.padding_idx();
    }
  }

  TORCH_CHECK(options.max_norm() == c10::nullopt)

  Tensor weight;
  scheduler.impl()->submit(
      Task{std::make_shared<Impl>(Impl{{weight}, options})});

  return std::make_shared<State>(
      State::Forward{std::move(weight)}, State::Backward{},
      State::Args{options.num_embeddings(), padding_idx, options.max_norm(),
                  options.norm_type(), options.scale_grad_by_freq(),
                  options.sparse()});
}

Tensor Embedding::forward(const Scheduler &scheduler,
                          const std::shared_ptr<State> &state,
                          const ReadOnlyTensor &indices) {
  struct Impl : Task::Impl {
    State::Args args;

    explicit Impl(std::vector<Tensor> output /* output */,
                  std::vector<ReadOnlyTensor> input /* weight, indices */,
                  const State::Args &args)
        : Task::Impl{std::move(output), std::move(input), kMain, kCompute},
          args{args} {}
    void operator()() const override {
      output()[0].impl()->tensor() = torch::embedding(
          input()[0].impl()->tensor(),
          input()[1].impl()->tensor().view(
              {-1, input()[1].impl()->tensor().size(-1)}),
          args.padding_idx, args.scale_grad_by_freq, args.sparse);
    }
    [[nodiscard]] const char *name() const override {
      return "cs::compute::Embedding::forward";
    }
  };

  Tensor output{};
  state->backward.indices = indices;
  scheduler.impl()->submit(Task{std::make_shared<Impl>(
      Impl{{output}, {state->forward.weight, indices}, state->args})});
  return output;
}

void Embedding::backward(const Scheduler &scheduler,
                         const std::shared_ptr<State> &state,
                         const ReadOnlyTensor &grad_output) {
  struct Impl : Task::Impl {
    State::Args args;

    explicit Impl(std::vector<Tensor> output /* grad_weight */,
                  std::vector<ReadOnlyTensor> input /* grad_output, indices */,
                  const State::Args &args)
        : Task::Impl{std::move(output), std::move(input), kAssist, kCompute},
          args{args} {}
    void operator()() const override {
      if (output()[0].impl()->tensor().defined()) {
        output()[0].impl()->tensor() += at::embedding_backward(
            input()[0].impl()->tensor(), input()[1].impl()->tensor(),
            args.num_weights, args.padding_idx, args.scale_grad_by_freq,
            args.sparse);
      } else /* accumulate grad */ {
        output()[0].impl()->tensor() = at::embedding_backward(
            input()[0].impl()->tensor(), input()[1].impl()->tensor(),
            args.num_weights, args.padding_idx, args.scale_grad_by_freq,
            args.sparse);
      }
    }
    [[nodiscard]] const char *name() const override {
      return "cs::compute::Embedding::backward";
    }
  };

  scheduler.impl()->submit(
      Task{std::make_shared<Impl>(Impl{{state->forward.grad_weight},
                                       {grad_output, state->backward.indices},
                                       state->args})});
  // decrease counter
  state->backward.indices.reset();
}
}  // namespace cs::compute
